File: java/src/main/java/net/snowflake/floe/AeadIv.java
package net.snowflake.floe;

import java.nio.ByteBuffer;
import java.security.SecureRandom;

class AeadIv {
  private final byte[] bytes;

  AeadIv(byte[] bytes) {
    this.bytes = bytes;
  }

  static AeadIv generateRandom(SecureRandom random, int ivLength) {
    byte[] iv = new byte[ivLength];
    random.nextBytes(iv);
    return new AeadIv(iv);
  }

  static AeadIv from(ByteBuffer buffer, int ivLength) {
    byte[] bytes = new byte[ivLength];
    buffer.get(bytes);
    return new AeadIv(bytes);
  }

  byte[] getBytes() {
    return bytes;
  }
}

File: cpp/src/AeadIv.cpp
#include "AeadIv.hpp"
#include "floe/FloeException.hpp"
#include <openssl/rand.h>

namespace floe {

AeadIv::AeadIv(const std::vector<uint8_t>& bytes) : bytes_(bytes) {
}

AeadIv AeadIv::generateRandom(const size_t ivLength) {
    std::vector<uint8_t> iv(ivLength);
    if (RAND_bytes(iv.data(), static_cast<int>(ivLength)) != 1) {
        throw FloeException("Failed to generate random IV");
    }
    return AeadIv(iv);
}

AeadIv AeadIv::from(const uint8_t* data, const size_t ivLength) {
    const std::vector bytes(data, data + ivLength);
    return AeadIv(bytes);
}

const std::vector<uint8_t>& AeadIv::getBytes() const {
    return bytes_;
}

}
